File: android/app/src/main/kotlin/com/uwbcompass/app/ranging/BleRangingProvider.kt
package com.uwbcompass.app.ranging

import com.uwbcompass.core.BleDistanceModel
import com.uwbcompass.core.RangingProvider
import com.uwbcompass.core.RangingSample
import com.uwbcompass.core.Technology
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

/**
 * BLE proximity provider: estimates DISTANCE ONLY from RSSI (no direction — ADR-0003).
 * The UI must render a proximity indicator (hot/cold), never an arrow, for BLE samples.
 *
 * The raw scan wiring (BluetoothLeScanner + advertising the session id) is device-only
 * and provided as an [RssiSource]; the RSSI→distance model and sample mapping below are
 * pure and unit-testable.
 */
class BleRangingProvider(
    private val rssiSource: RssiSource,
    private val txPowerDbm: Int = -59, // measured RSSI at 1 m; VERIFY-ON-DEVICE per phone.
    private val pathLossExponent: Double = 2.0,
) : RangingProvider {

    fun interface RssiSource {
        /** Cold flow of (rssiDbm, timestampMs). Collecting starts the BLE scan. */
        fun rssi(): Flow<Pair<Int, Long>>
    }

    override val technology = Technology.BLE

    override fun samples(): Flow<RangingSample> = rssiSource.rssi().map { (rssi, ts) ->
        val distance = BleDistanceModel.rssiToDistanceMeters(rssi, txPowerDbm, pathLossExponent)
        RangingSample(
            technology = Technology.BLE,
            distanceMeters = distance,
            azimuthDeg = null,
            elevationDeg = null,
            quality = BleDistanceModel.qualityFrom(distance),
            timestampMs = ts,
        )
    }
}

File: ios/App/Ranging/BleRangingProvider.swift
import Foundation
import UWBCompassCore
#if canImport(CoreBluetooth)
import CoreBluetooth
#endif

/// BLE proximity provider: DISTANCE ONLY from RSSI (no direction — ADR-0003). Uses the
/// shared `BleDistanceModel` (unit-tested in the core package). CoreBluetooth scanning is
/// device-only; the RSSI→distance mapping is shared and verified.
final class BleRangingProvider: NSObject, RangingProvider {
    let technology: Technology = .BLE

    /// Measured RSSI at 1 m for the paired hardware (VERIFY-ON-DEVICE calibration).
    private let txPowerDbm: Int
    private let serviceUUIDString: String
    private var continuation: AsyncStream<RangingSample>.Continuation?

    init(serviceUUIDString: String, txPowerDbm: Int = -59) {
        self.serviceUUIDString = serviceUUIDString
        self.txPowerDbm = txPowerDbm
    }

    func samples() -> AsyncStream<RangingSample> {
        AsyncStream { continuation in
            self.continuation = continuation
            #if canImport(CoreBluetooth)
            // VERIFY-ON-DEVICE: CBCentralManager scan for serviceUUIDString; each
            // didDiscover RSSI -> emit(rssi:). Advertise our own session UUID via CBPeripheralManager.
            #else
            continuation.finish()
            #endif
        }
    }

    /// Emit a sample for an observed RSSI. Shared, testable path.
    func emit(rssi: Int, timestampMs: Int64) {
        let distance = BleDistanceModel.rssiToDistanceMeters(rssi: rssi, txPowerDbm: txPowerDbm)
        continuation?.yield(RangingSample(
            technology: .BLE, distanceMeters: distance, azimuthDeg: nil, elevationDeg: nil,
            quality: BleDistanceModel.quality(forDistance: distance), timestampMs: timestampMs
        ))
    }
}
